Bug: Intermittent failure in keyvault controller test case

**Describe the bug**
A clear and concise description of what the bug is.

The following error occurs intermittently in the keyvalue_controller_test, though not sure if it is directly related
.
There appears to be an error condition that appears while logging from the following line:

```
	/home/vsts/work/1/s/gopath/src/Azure/azure-service-operator/controllers/keyvault_controller.go:82 +0xb34
```
(line 80-85 `/controllers/keyvault_controller.go` listed below):
```
	if !helpers.HasFinalizer(&instance, keyVaultFinalizerName) {
		if err := r.addFinalizer(&instance); err != nil {
			log.Info("Adding keyvault finalizer failed with ", err.Error())
			return ctrl.Result{}, err
		}
	}
```

though the cause may be unrelated.

An example log is the following:
https://dev.azure.com/azure/azure-service-operator/_build/results?buildId=3655&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9

```
panic: odd number of arguments passed as key-value pairs for logging [recovered]
	panic: odd number of arguments passed as key-value pairs for logging

goroutine 396 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/vsts/work/1/s/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:58 +0x105
panic(0x154dee0, 0xc000572500)
	/usr/local/go1.12/src/runtime/panic.go:522 +0x1b5
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc0000b0580, 0xc000088fc0, 0x1, 0x1)
	/home/vsts/work/1/s/gopath/pkg/mod/go.uber.org/zap@v1.9.1/zapcore/entry.go:229 +0x546
go.uber.org/zap.(*Logger).DPanic(0xc00045f020, 0x1801361, 0x3d, 0xc000088fc0, 0x1, 0x1)
	/home/vsts/work/1/s/gopath/pkg/mod/go.uber.org/zap@v1.9.1/logger.go:215 +0x7f
github.com/go-logr/zapr.handleFields(0xc00045f020, 0xc0005724b0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x10, 0x155d320, 0x1)
	/home/vsts/work/1/s/gopath/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:106 +0x5ce
github.com/go-logr/zapr.(*infoLogger).Info(0xc0007dbf48, 0x17d6cc7, 0x26, 0xc0005724b0, 0x1, 0x1)
	/home/vsts/work/1/s/gopath/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:70 +0xb2
github.com/Azure/azure-service-operator/controllers.(*KeyVaultReconciler).Reconcile(0xc0005b5100, 0xc00057ab49, 0x7, 0xc000680920, 0x13, 0x2878dc0, 0xc000948900, 0xc000357588, 0xc0003575b8)
	/home/vsts/work/1/s/gopath/src/Azure/azure-service-operator/controllers/keyvault_controller.go:82 +0xb34
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0003300a0, 0x162eb40, 0xc0007dbec0, 0x162eb00)
	/home/vsts/work/1/s/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.4/pkg/internal/controller/controller.go:216 +0x146
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0003300a0, 0xc000293c00)
	/home/vsts/work/1/s/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.4/pkg/internal/controller/controller.go:192 +0xb5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0003300a0)
	/home/vsts/work/1/s/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.4/pkg/internal/controller/controller.go:171 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000396900)
	/home/vsts/work/1/s/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000396900, 0x3b9aca00, 0x0, 0x1, 0xc0004cecc0)
	/home/vsts/work/1/s/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000396900, 0x3b9aca00, 0xc0004cecc0)
	/home/vsts/work/1/s/gopath/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/home/vsts/work/1/s/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.4/pkg/internal/controller/controller.go:157 +0x311
FAIL	github.com/Azure/azure-service-operator/controllers	145.749s
```
**To Reproduce**
Steps to reproduce the behavior:

Difficult to reproduce, as error occurs intermittently.

**Expected behavior**

As described above.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/.devcontainer/devcontainer.json
+++ b/.devcontainer/devcontainer.json
@@ -1,24 +1,25 @@
 // If you want to run as a non-root user in the container, see .devcontainer/docker-compose.yml.
 {
-	"name": "Go",
-	"dockerComposeFile": "docker-compose.yml",
+    "name": "Go",
+    "dockerComposeFile": "docker-compose.yml",
     "service": "docker-in-docker",
     "workspaceFolder": "/workspace",
     "shutdownAction": "stopCompose",
-	"extensions": [
-		"ms-azuretools.vscode-docker",
-		"ms-vscode.go"
-	],
-	"settings": {
-		"terminal.integrated.shell.linux": "zsh",
-		"go.gopath": "/go",
-		"go.inferGopath": true,
-		"go.useLanguageServer": true,
-		"go.toolsEnvVars": {
-			"GO111MODULE": "on"
-		},
-		"remote.extensionKind": {
-			"ms-azuretools.vscode-docker": "workspace"
-		}
-	}
+    "postCreateCommand": "make set-kindcluster",
+    "extensions": [
+        "ms-azuretools.vscode-docker",
+        "ms-vscode.go"
+    ],
+    "settings": {
+        "terminal.integrated.shell.linux": "zsh",
+        "go.gopath": "/go",
+        "go.inferGopath": true,
+        "go.useLanguageServer": true,
+        "go.toolsEnvVars": {
+            "GO111MODULE": "on"
+        },
+        "remote.extensionKind": {
+            "ms-azuretools.vscode-docker": "workspace"
+        }
+    }
 }

--- a/.devcontainer/docker-compose.yml
+++ b/.devcontainer/docker-compose.yml
@@ -1,7 +1,7 @@
 version: '3'
 services:
   docker-in-docker:
-    build: 
+    build:
       context: ../
       dockerfile: .devcontainer/Dockerfile
     network_mode: "host"
@@ -15,8 +15,8 @@ services:
       # Forwarding the socket is optional, but lets docker work inside the container if you install the Docker CLI.
       # See the docker-in-docker-compose definition for details on how to install it.
       - /var/run/docker.sock:/var/run/docker.sock
-    
+
     # Overrides default command so things don't shut down after the process ends - useful for debugging
-    command: bash -c "cd /workspace/ && make set-kindcluster && sleep infinity"
+    command: sleep infinity
     env_file:
       - .env

--- a/Makefile
+++ b/Makefile
@@ -51,9 +51,11 @@ deploy: manifests
 	kubectl apply -f config/crd/bases
 	kustomize build config/default | kubectl apply -f -
 
+timestamp := $(shell /bin/date "+%Y%m%d-%H%M%S")
+
 update:
-	IMG="docker.io/controllertest:1" make ARGS="${ARGS}" docker-build
-	kind load docker-image docker.io/controllertest:1 --loglevel "trace"
+	IMG="docker.io/controllertest:$(timestamp)" make ARGS="${ARGS}" docker-build
+	kind load docker-image docker.io/controllertest:$(timestamp) --loglevel "trace"
 	make install
 	make deploy
 	sed -i'' -e 's@image: .*@image: '"IMAGE_URL"'@' ./config/default/manager_image_patch.yaml

--- a/README.md
+++ b/README.md
@@ -46,7 +46,7 @@ Basic commands to check your cluster
 
 ### Quick Start
 
-If you're using VSCode with [Remote - Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extensions installed, you quickly have you're environment set up and ready to go with everything you need to get started.
+If you're using VSCode with [Remote - Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extensions installed, you can quickly have you're environment set up and ready to go with everything you need to get started.
 
 1. Open this project in VSCode.
 2. Inside `.devcontainer`, create a file called `.env` and using the following template, copy your Service Principal's details.
@@ -65,9 +65,7 @@ If you're using VSCode with [Remote - Containers](https://marketplace.visualstud
 4. VSCode will relaunch and start building our development container. This will install all the necessary dependencies required for you to begin developing.
 5. Once the container has finished building, you can now start testing your Azure Service Operator within your own local kubernetes environment.
 
-**Note**: after the DevContainer has finished building, the kind cluster will initialising and installing the Azure Service Operator in the background. This will take some time before it is available.
-
-To see when the kind cluster is ready, use `docker ps -a` to list your running containers, look for `IMAGE` with the name `azure-service-operator_devcontainer_docker-in-docker...`. Using that image's `CONTAINER ID`, use `docker logs -f CONTAINER ID` to view the logs from the container setting up your cluster.
+**Note**: if you do not want to create a kind cluster when starting the devcontainer, comment out `"postCreateCommand": "make set-kindcluster",` in `.devcontainer/devcontainer.json` and reopen the devcontainer.
 
 ### Getting started
 

--- a/config/samples/azure_v1_storage.yaml
+++ b/config/samples/azure_v1_storage.yaml
@@ -1,7 +1,7 @@
 apiVersion: azure.microsoft.com/v1
 kind: Storage
 metadata:
-  name: storagesample1908xyzkj
+  name: storagesample1908ayzkj
 spec:
   location: westus
   resourceGroup: resourcegroup-sample-1908

--- a/controllers/cosmosdb_controller.go
+++ b/controllers/cosmosdb_controller.go
@@ -69,14 +69,12 @@ func (r *CosmosDBReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 
 	if err := r.Get(ctx, req.NamespacedName, &instance); err != nil {
-		log.Error(err, "unable to fetch CosmosDB")
+		log.Info("Unable to retrieve cosmosDB resource", "err", err.Error())
 		// we'll ignore not-found errors, since they can't be fixed by an immediate
 		// requeue (we'll need to wait for a new notification), and we can get them
 		// on deleted requests.
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
-	log.Info("Getting CosmosDB Account", "CosmosDB.Namespace", instance.Namespace, "CosmosDB.Name", instance.Name)
-	log.V(1).Info("Describing CosmosDB Account", "CosmosDB", instance)
 
 	if helpers.IsBeingDeleted(&instance) {
 		if helpers.HasFinalizer(&instance, cosmosDBFinalizerName) {
@@ -94,27 +92,28 @@ func (r *CosmosDBReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 
 	if !helpers.HasFinalizer(&instance, cosmosDBFinalizerName) {
-		if err := r.addFinalizer(&instance); err != nil {
-			log.Info("Error", "Adding cosmosDB finalizer failed with ", err)
-			return ctrl.Result{}, err
-		}
-	}
-
-	if !instance.IsSubmitted() {
-		if err := r.reconcileExternal(&instance); err != nil {
-			if errhelp.IsAsynchronousOperationNotComplete(err) || errhelp.IsGroupNotFound(err) {
-				log.Info("Requeuing as the async operation is not complete")
-				return ctrl.Result{
-					Requeue:      true,
-					RequeueAfter: time.Second * time.Duration(requeueAfter),
-				}, nil
-			}
-			return ctrl.Result{}, fmt.Errorf("error reconciling cosmosdb in azure: %v", err)
+		err := r.addFinalizer(&instance)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error when adding finalizer: %v", err)
 		}
 		return ctrl.Result{}, nil
 	}
 
-	r.Recorder.Event(&instance, "Normal", "Provisioned", "CosmosDB "+instance.ObjectMeta.Name+" provisioned ")
+	if !instance.IsSubmitted() {
+		err := r.reconcileExternal(&instance)
+		if err != nil {
+			catch := []string{
+				errhelp.ParentNotFoundErrorCode,
+				errhelp.ResourceGroupNotFoundErrorCode,
+			}
+			if helpers.ContainsString(catch, err.(*errhelp.AzureError).Type) {
+				log.Info("Got ignorable error", "type", err.(*errhelp.AzureError).Type)
+				return ctrl.Result{Requeue: true, RequeueAfter: time.Second * time.Duration(requeueAfter)}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("error when creating resource in azure: %v", err)
+		}
+		return ctrl.Result{}, nil
+	}
 	return ctrl.Result{}, nil
 }
 
@@ -136,34 +135,34 @@ func (r *CosmosDBReconciler) reconcileExternal(instance *azurev1.CosmosDB) error
 	kind := instance.Spec.Kind
 	dbType := instance.Spec.Properties.DatabaseAccountOfferType
 
+	var err error
+
 	// write information back to instance
 	instance.Status.Provisioning = true
-
-	if err := r.Status().Update(ctx, instance); err != nil {
+	err = r.Update(ctx, instance)
+	if err != nil {
+		//log error and kill it
 		r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 	}
-
-	_, err := cosmosdbs.CreateCosmosDB(ctx, groupName, name, location, kind, dbType, nil)
+	_, err = cosmosdbs.CreateCosmosDB(ctx, groupName, name, location, kind, dbType, nil)
 	if err != nil {
-		if errhelp.IsAsynchronousOperationNotComplete(err) || errhelp.IsGroupNotFound(err) {
-			r.Recorder.Event(instance, "Normal", "Provisioning", name+" provisioning")
-			return err
-		}
 		r.Recorder.Event(instance, "Warning", "Failed", "Couldn't create resource in azure")
 		instance.Status.Provisioning = false
-		errUpdate := r.Status().Update(ctx, instance)
+		errUpdate := r.Update(ctx, instance)
 		if errUpdate != nil {
+			//log error and kill it
 			r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 		}
-		return err
+		return errhelp.NewAzureError(err)
 	}
-
 	instance.Status.Provisioning = false
 	instance.Status.Provisioned = true
 
-	if err = r.Status().Update(ctx, instance); err != nil {
+	err = r.Update(ctx, instance)
+	if err != nil {
 		r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 	}
+	r.Recorder.Event(instance, "Normal", "Updated", name+" provisioned")
 
 	return nil
 }

--- a/controllers/rediscache_controller.go
+++ b/controllers/rediscache_controller.go
@@ -68,14 +68,12 @@ func (r *RedisCacheReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) 
 	}
 
 	if err := r.Get(ctx, req.NamespacedName, &instance); err != nil {
-		log.Error(err, "unable to fetch RedisCache")
+		log.Info("Unable to retrieve redis cache resource", "err", err.Error())
 		// we'll ignore not-found errors, since they can't be fixed by an immediate
 		// requeue (we'll need to wait for a new notification), and we can get them
 		// on deleted requests.
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
-	log.Info("Getting Redis Cache", "RedisCache.Namespace", instance.Namespace, "RedisCache.Name", instance.Name)
-	log.V(1).Info("Describing Redis Cache", "RedisCache", instance)
 
 	if helpers.IsBeingDeleted(&instance) {
 		if helpers.HasFinalizer(&instance, redisCacheFinalizerName) {
@@ -93,27 +91,28 @@ func (r *RedisCacheReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) 
 	}
 
 	if !helpers.HasFinalizer(&instance, redisCacheFinalizerName) {
-		if err := r.addFinalizer(&instance); err != nil {
-			log.Info("Error", "Adding redis cache finalizer failed with ", err)
-			return ctrl.Result{}, err
-		}
-	}
-
-	if !instance.IsSubmitted() {
-		if err := r.reconcileExternal(&instance); err != nil {
-			if errhelp.IsAsynchronousOperationNotComplete(err) || errhelp.IsGroupNotFound(err) {
-				log.Info("Requeuing as the async operation is not complete")
-				return ctrl.Result{
-					Requeue:      true,
-					RequeueAfter: time.Second * time.Duration(requeueAfter),
-				}, nil
-			}
-			return ctrl.Result{}, fmt.Errorf("error reconciling redis cache in azure: %v", err)
+		err := r.addFinalizer(&instance)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error when adding finalizer: %v", err)
 		}
 		return ctrl.Result{}, nil
 	}
 
-	r.Recorder.Event(&instance, "Normal", "Provisioned", "RedisCache "+instance.ObjectMeta.Name+" provisioned ")
+	if !instance.IsSubmitted() {
+		err := r.reconcileExternal(&instance)
+		if err != nil {
+			catch := []string{
+				errhelp.ParentNotFoundErrorCode,
+				errhelp.ResourceGroupNotFoundErrorCode,
+			}
+			if helpers.ContainsString(catch, err.(*errhelp.AzureError).Type) {
+				log.Info("Got ignorable error", "type", err.(*errhelp.AzureError).Type)
+				return ctrl.Result{Requeue: true, RequeueAfter: time.Second * time.Duration(requeueAfter)}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("error when creating resource in azure: %v", err)
+		}
+		return ctrl.Result{}, nil
+	}
 	return ctrl.Result{}, nil
 }
 
@@ -129,6 +128,9 @@ func (r *RedisCacheReconciler) addFinalizer(instance *azurev1.RedisCache) error 
 
 func (r *RedisCacheReconciler) reconcileExternal(instance *azurev1.RedisCache) error {
 	ctx := context.Background()
+
+	var err error
+
 	location := instance.Spec.Location
 	name := instance.ObjectMeta.Name
 	groupName := instance.Spec.ResourceGroupName
@@ -138,31 +140,33 @@ func (r *RedisCacheReconciler) reconcileExternal(instance *azurev1.RedisCache) e
 	// write information back to instance
 	instance.Status.Provisioning = true
 
-	if err := r.Status().Update(ctx, instance); err != nil {
+	err = r.Update(ctx, instance)
+	if err != nil {
+		//log error and kill it
 		r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 	}
 
-	_, err := rediscaches.CreateRedisCache(ctx, groupName, name, location, sku, enableNonSSLPort, nil)
+	_, err = rediscaches.CreateRedisCache(ctx, groupName, name, location, sku, enableNonSSLPort, nil)
 	if err != nil {
-		if errhelp.IsAsynchronousOperationNotComplete(err) || errhelp.IsGroupNotFound(err) {
-			r.Recorder.Event(instance, "Normal", "Provisioning", name+" provisioning")
-			return err
-		}
 		r.Recorder.Event(instance, "Warning", "Failed", "Couldn't create resource in azure")
 		instance.Status.Provisioning = false
-		errUpdate := r.Status().Update(ctx, instance)
+		errUpdate := r.Update(ctx, instance)
 		if errUpdate != nil {
+			//log error and kill it
 			r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 		}
-		return err
+		return errhelp.NewAzureError(err)
 	}
 
 	instance.Status.Provisioning = false
 	instance.Status.Provisioned = true
 
-	if err = r.Status().Update(ctx, instance); err != nil {
+	err = r.Update(ctx, instance)
+	if err != nil {
 		r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 	}
+
+	r.Recorder.Event(instance, "Normal", "Updated", name+" provisioned")
 
 	return nil
 }
@@ -181,7 +185,6 @@ func (r *RedisCacheReconciler) deleteExternal(instance *azurev1.RedisCache) erro
 		r.Recorder.Event(instance, "Warning", "Failed", "Couldn't delete resouce in azure")
 		return err
 	}
-
 	r.Recorder.Event(instance, "Normal", "Deleted", name+" deleted")
 	return nil
 }

--- a/controllers/storage_controller.go
+++ b/controllers/storage_controller.go
@@ -37,7 +37,7 @@ import (
 
 	azurev1 "github.com/Azure/azure-service-operator/api/v1"
 	"github.com/Azure/azure-service-operator/pkg/errhelp"
-	helpers "github.com/Azure/azure-service-operator/pkg/helpers"
+	"github.com/Azure/azure-service-operator/pkg/helpers"
 	"github.com/Azure/azure-service-operator/pkg/resourcemanager/storages"
 	"k8s.io/client-go/tools/record"
 )
@@ -69,7 +69,7 @@ func (r *StorageReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 
 	if err := r.Get(ctx, req.NamespacedName, &instance); err != nil {
-		log.Error(err, "unable to retrieve storage resource", "err", err.Error())
+		log.Info("unable to retrieve storage resource", "err", err.Error())
 		// we'll ignore not-found errors, since they can't be fixed by an immediate
 		// requeue (we'll need to wait for a new notification), and we can get them
 		// on deleted requests.
@@ -77,8 +77,8 @@ func (r *StorageReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	log.Info("Getting Storage Account", "Storage.Namespace", instance.Namespace, "Storage.Name", instance.Name)
-	log.V(1).Info("Describing Storage Account", "Storage", instance)
+	//log.Info("Getting Storage Account", "Storage.Namespace", instance.Namespace, "Storage.Name", instance.Name)
+	//log.V(1).Info("Describing Storage Account", "Storage", instance)
 
 	if helpers.IsBeingDeleted(&instance) {
 		if helpers.HasFinalizer(&instance, storageFinalizerName) {
@@ -96,27 +96,29 @@ func (r *StorageReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 
 	if !helpers.HasFinalizer(&instance, storageFinalizerName) {
-		if err := r.addFinalizer(&instance); err != nil {
-			log.Info("Error", "Adding storage finalizer failed with ", err)
-			return ctrl.Result{}, err
-		}
-	}
-
-	if !instance.IsSubmitted() {
-		if err := r.reconcileExternal(&instance); err != nil {
-			if errhelp.IsAsynchronousOperationNotComplete(err) || errhelp.IsGroupNotFound(err) {
-				log.Info("Requeuing as the async operation is not complete")
-				return ctrl.Result{
-					Requeue:      true,
-					RequeueAfter: time.Second * time.Duration(requeueAfter),
-				}, nil
-			}
-			return ctrl.Result{}, fmt.Errorf("error reconciling storage in azure: %v", err)
+		err := r.addFinalizer(&instance)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error when adding finalizer: %v", err)
 		}
 		return ctrl.Result{}, nil
 	}
 
-	r.Recorder.Event(&instance, "Normal", "Provisioned", "Storage "+instance.ObjectMeta.Name+" provisioned ")
+	if !instance.IsSubmitted() {
+		err := r.reconcileExternal(&instance)
+		if err != nil {
+			catch := []string{
+				errhelp.ParentNotFoundErrorCode,
+				errhelp.ResourceGroupNotFoundErrorCode,
+			}
+			if helpers.ContainsString(catch, err.(*errhelp.AzureError).Type) {
+				log.Info("Got ignorable error", "type", err.(*errhelp.AzureError).Type)
+				return ctrl.Result{Requeue: true, RequeueAfter: time.Second * time.Duration(requeueAfter)}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("error when creating resource in azure: %v", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
 	return ctrl.Result{}, nil
 }
 
@@ -140,34 +142,38 @@ func (r *StorageReconciler) reconcileExternal(instance *azurev1.Storage) error {
 	accessTier := instance.Spec.AccessTier
 	enableHTTPSTrafficOnly := instance.Spec.EnableHTTPSTrafficOnly
 
+	var err error
+
 	// write information back to instance
 	instance.Status.Provisioning = true
 
-	if err := r.Status().Update(ctx, instance); err != nil {
+	err = r.Update(ctx, instance)
+	if err != nil {
+		//log error and kill it
 		r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 	}
 
-	_, err := storages.CreateStorage(ctx, groupName, name, location, sku, kind, nil, accessTier, enableHTTPSTrafficOnly)
+	_, err = storages.CreateStorage(ctx, groupName, name, location, sku, kind, nil, accessTier, enableHTTPSTrafficOnly)
 	if err != nil {
-		if errhelp.IsAsynchronousOperationNotComplete(err) || errhelp.IsGroupNotFound(err) {
-			r.Recorder.Event(instance, "Normal", "Provisioning", name+" provisioning")
-			return err
-		}
 		r.Recorder.Event(instance, "Warning", "Failed", "Couldn't create resource in azure")
 		instance.Status.Provisioning = false
-		errUpdate := r.Status().Update(ctx, instance)
+		errUpdate := r.Update(ctx, instance)
 		if errUpdate != nil {
+			//log error and kill it
 			r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 		}
-		return err
+		return errhelp.NewAzureError(err)
 	}
 
 	instance.Status.Provisioning = false
 	instance.Status.Provisioned = true
 
-	if err = r.Status().Update(ctx, instance); err != nil {
+	err = r.Update(ctx, instance)
+	if err != nil {
 		r.Recorder.Event(instance, "Warning", "Failed", "Unable to update instance")
 	}
+
+	r.Recorder.Event(instance, "Normal", "Updated", name+" provisioned")
 
 	return nil
 }

--- a/pkg/resourcemanager/cosmosdbs/cosmosdbs.go
+++ b/pkg/resourcemanager/cosmosdbs/cosmosdbs.go
@@ -1,3 +1,27 @@
+/*
+MIT License
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE
+*/
+
 package cosmosdbs
 
 import (
@@ -29,22 +53,8 @@ func CreateCosmosDB(ctx context.Context, groupName string,
 	location string,
 	kind azurev1.CosmosDBKind,
 	dbType azurev1.CosmosDBDatabaseAccountOfferType,
-	tags map[string]*string) (documentdb.DatabaseAccount, error) {
+	tags map[string]*string) (*documentdb.DatabaseAccount, error) {
 	cosmosDBClient := getCosmosDBClient()
-
-	log.Println("CosmosDB:CosmosDBName" + cosmosDBName)
-
-	/* Uncomment and update if we should be checking for name exists first
-	result, err = cosmosDBClient.CheckNameExists(ctx, cosmosDBName)
-	if err != nil {
-		return documentdb.DatabaseAccount.{}, err
-	}
-	result.
-	if *result.NameAvailable == false {
-		log.Fatalf("storage account not available: %v\n", result.Reason)
-		return storage.Account{}, errors.New("storage account not available")
-	}*/
-
 	dbKind := documentdb.DatabaseAccountKind(kind)
 	sDBType := string(dbType)
 
@@ -70,7 +80,6 @@ func CreateCosmosDB(ctx context.Context, groupName string,
 	locationsArray := []documentdb.Location{
 		locationObj,
 	}
-
 	createUpdateParams := documentdb.DatabaseAccountCreateUpdateParameters{
 		Location: to.StringPtr(location),
 		Tags:     tags,
@@ -85,16 +94,20 @@ func CreateCosmosDB(ctx context.Context, groupName string,
 			Locations:                     &locationsArray,
 		},
 	}
-
-	log.Println(fmt.Sprintf("creating cosmosDB '%s' in resource group '%s' and location: %v", cosmosDBName, groupName, location))
-
 	future, err := cosmosDBClient.CreateOrUpdate(
 		ctx, groupName, cosmosDBName, createUpdateParams)
+
 	if err != nil {
-		log.Println(fmt.Sprintf("ERROR creating cosmosDB '%s' in resource group '%s' and location: %v", cosmosDBName, groupName, location))
-		log.Println(fmt.Printf("failed to initialize cosmosdb: %v\n", err))
+		return nil, err
 	}
-	return future.Result(cosmosDBClient)
+
+	err = future.WaitForCompletionRef(ctx, cosmosDBClient.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := future.Result(cosmosDBClient)
+	return &result, err
 }
 
 // DeleteCosmosDB removes the resource group named by env var

--- a/pkg/resourcemanager/rediscaches/rediscaches.go
+++ b/pkg/resourcemanager/rediscaches/rediscaches.go
@@ -1,9 +1,32 @@
+/*
+MIT License
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE
+*/
+
 package rediscaches
 
 import (
 	"context"
 	"errors"
-	"fmt"
 	"log"
 
 	"github.com/Azure/azure-sdk-for-go/services/redis/mgmt/2018-03-01/redis"
@@ -31,10 +54,8 @@ func CreateRedisCache(ctx context.Context,
 	location string,
 	sku azurev1.RedisCacheSku,
 	enableNonSSLPort bool,
-	tags map[string]*string) (redis.ResourceType, error) {
+	tags map[string]*string) (*redis.ResourceType, error) {
 	redisClient := getRedisCacheClient()
-
-	log.Println("RedisCache:CacheName" + redisCacheName)
 
 	//Check if name is available
 	redisType := "Microsoft.Cache/redis"
@@ -42,17 +63,15 @@ func CreateRedisCache(ctx context.Context,
 		Name: &redisCacheName,
 		Type: &redisType,
 	}
-	result, err := redisClient.CheckNameAvailability(ctx, checkNameParams)
+	checkNameResult, err := redisClient.CheckNameAvailability(ctx, checkNameParams)
 	if err != nil {
-		return redis.ResourceType{}, err
+		return nil, err
 	}
 
-	if result.StatusCode != 200 {
-		log.Fatalf("redis cache name (%s) not available: %v\n", redisCacheName, result.Status)
-		return redis.ResourceType{}, errors.New("redis cache name not available")
+	if checkNameResult.StatusCode != 200 {
+		log.Fatalf("redis cache name (%s) not available: %v\n", redisCacheName, checkNameResult.Status)
+		return nil, errors.New("redis cache name not available")
 	}
-
-	log.Println(fmt.Sprintf("creating rediscache '%s' in resource group '%s' and location: %v", redisCacheName, groupName, location))
 
 	redisSku := redis.Sku{
 		Name:     redis.SkuName(sku.Name),
@@ -72,10 +91,15 @@ func CreateRedisCache(ctx context.Context,
 	future, err := redisClient.Create(
 		ctx, groupName, redisCacheName, createParams)
 	if err != nil {
-		log.Println(fmt.Sprintf("ERROR creating redisCache '%s' in resource group '%s' and location: %v", redisCacheName, groupName, location))
-		log.Println(fmt.Printf("failed to initialize redis Cache: %v\n", err))
+		return nil, err
 	}
-	return future.Result(redisClient)
+
+	err = future.WaitForCompletionRef(ctx, redisClient.Client)
+	if err != nil {
+		return nil, err
+	}
+	result, err := future.Result(redisClient)
+	return &result, err
 }
 
 // DeleteRedisCache removes the resource group named by env var

--- a/pkg/resourcemanager/storages/storages.go
+++ b/pkg/resourcemanager/storages/storages.go
@@ -1,9 +1,32 @@
+/*
+MIT License
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE
+*/
+
 package storages
 
 import (
 	"context"
 	"errors"
-	"fmt"
 	"log"
 
 	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
@@ -33,19 +56,20 @@ func CreateStorage(ctx context.Context, groupName string,
 	kind apiv1.StorageKind,
 	tags map[string]*string,
 	accessTier apiv1.StorageAccessTier,
-	enableHTTPsTrafficOnly *bool) (storage.Account, error) {
+	enableHTTPsTrafficOnly *bool) (*storage.Account, error) {
 	storagesClient := getStoragesClient()
 
+	//Check if name is available
 	storageType := "Microsoft.Storage/storageAccounts"
 	checkAccountParams := storage.AccountCheckNameAvailabilityParameters{Name: &storageAccountName, Type: &storageType}
-	result, err := storagesClient.CheckNameAvailability(ctx, checkAccountParams)
+	checkNameResult, err := storagesClient.CheckNameAvailability(ctx, checkAccountParams)
 	if err != nil {
-		return storage.Account{}, err
+		return nil, err
 	}
 
-	if *result.NameAvailable == false {
-		log.Fatalf("storage account not available: %v\n", result.Reason)
-		return storage.Account{}, errors.New("storage account not available")
+	if *checkNameResult.NameAvailable == false {
+		log.Fatalf("storage account not available: %v\n", checkNameResult.Reason)
+		return nil, errors.New("storage account name not available")
 	}
 
 	sSku := storage.Sku{Name: storage.SkuName(sku.Name)}
@@ -64,15 +88,18 @@ func CreateStorage(ctx context.Context, groupName string,
 		},
 	}
 
-	log.Println(fmt.Sprintf("creating storage '%s' in resource group '%s' and location: %v", storageAccountName, groupName, location))
+	//log.Println(fmt.Sprintf("creating storage '%s' in resource group '%s' and location: %v", storageAccountName, groupName, location))
 	future, err := storagesClient.Create(ctx, groupName, storageAccountName, params)
+	if err != nil {
+		return nil, err
+	}
 
 	err = future.WaitForCompletionRef(ctx, storagesClient.Client)
 	if err != nil {
-		return storage.Account{}, err
+		return nil, err
 	}
-
-	return future.Result(storagesClient)
+	result, err := future.Result(storagesClient)
+	return &result, err
 }
 
 // Get gets the description of the specified storage account.

